### PR TITLE
#1370: Introduce TransactionBuilder

### PR DESCRIPTION
**Pull request by andreoss**
Per #1370 
Refactoring towards TransactionBuilder

## Description
- Add TransactionBuilder, refactor users of Transaction to use the builder instead of direct call to one of contructors
- Remove constructors replacable by the builder

## Motivation and Context
- Tech debt & code quality

## How Has This Been Tested?-
- Unit-tests pass 

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Requires Activation Code (Hard Fork)

* **Other information**:

### CODE REVIEW

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java [hunk 1]
```diff
@@ -101,20 +100,16 @@
     private Keccak256 hash;
     private Keccak256 rawHash;
 
-    protected Transaction(byte[] rawData) {
-        RLPList transaction = RLP.decodeList(rawData);
-
+    protected Transaction(RLPList transaction) {
```
Comment (patogallaiovlabs) [inline rskj-core/src/main/java/org/ethereum/core/Transaction.java:103]: Why did you change the byte array for RLPList? whats the rationale?
And why didn't you change it in subclasses?

Comment (andreoss) [reply rskj-core/src/main/java/org/ethereum/core/Transaction.java:103]: @patogallaiovlabs I've tried to reduce amount of `byte[]` arrays passed around as arguments, which brings a stronger typing. 
It might be unnecessary here, as subclasses seem to be about of scope of this task.

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java [hunk 2]
```diff
@@ -137,56 +132,38 @@ protected Transaction(byte[] rawData) {
...
-    public Transaction(String to, BigInteger amount, BigInteger nonce, BigInteger gasPrice, BigInteger gasLimit, byte[] decodedData, byte chainId) {
-        this(BigIntegers.asUnsignedByteArray(nonce),
-                gasPrice.toByteArray(),
-                BigIntegers.asUnsignedByteArray(gasLimit),
-                to != null ? Hex.decode(to) : null,
-                BigIntegers.asUnsignedByteArray(amount),
-                decodedData,
-                chainId);
+        this(
+                ByteUtil.cloneBytes(nonce),
```
Comment (ajlopezrsk) [inline rskj-core/src/main/java/org/ethereum/core/Transaction.java:142]: Could we clone these fields in the innermost constructor?

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java [hunk 3]
```diff
@@ -137,56 +132,38 @@ protected Transaction(byte[] rawData) {
...
+                ByteUtil.cloneBytes(gasLimit),
+                RLP.parseRskAddress(ByteUtil.cloneBytes(receiveAddress)),
+                RLP.parseCoinNullZero(ByteUtil.cloneBytes(valueRaw)),
+                ByteUtil.cloneBytes(data),
+                chainId,
+                false
+        );
+    }
+
+    Transaction(byte[] nonce, Coin gasPriceRaw, byte[] gasLimit, RskAddress receiveAddress, Coin valueRaw, byte[] data,
```
Comment (ajlopezrsk) [inline rskj-core/src/main/java/org/ethereum/core/Transaction.java:153]: Any suggested explicit visibility?

File: rskj-core/src/main/java/co/rsk/remasc/RemascTransaction.java [hunk 4]
```diff
@@ -53,7 +54,7 @@ public String toJsonString() {
     };
 
     public RemascTransaction(byte[] rawData) {
-        super(rawData);
+        super(RLP.decodeList(rawData));
```
Comment (ajlopezrsk) [inline rskj-core/src/main/java/co/rsk/remasc/RemascTransaction.java:57]: Maybe, out of scope, but it could be nice to have RLP logic decoupled from Transaction class, at least in this constructor

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java (hunk 3 above)
Comment (andreoss) [reply rskj-core/src/main/java/org/ethereum/core/Transaction.java:153]: @ajlopezrsk Replaced implicit with `protected`.

File: rskj-core/src/main/java/co/rsk/remasc/RemascTransaction.java (hunk 4 above)
Comment (andreoss) [reply rskj-core/src/main/java/co/rsk/remasc/RemascTransaction.java:57]: @ajlopezrsk Now `RLP` logic is handled by the base class itself.

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java (hunk 2 above)
Comment (andreoss) [reply rskj-core/src/main/java/org/ethereum/core/Transaction.java:142]: @ajlopezrsk Done

**Comment by ajlopezrsk** [comment]
pipeline:run

**Review by ajlopezrsk — APPROVED** [review]

**Review by fedejinich — DISMISSED** [review]
lgtm

**Comment by ajlopezrsk** [comment]
@andreoss please, could you rebase your contribution to current master?

**Comment by andreoss** [comment]
Done.

**Review by ajlopezrsk — APPROVED** [review]

**Comment by alejandrod-iov** [comment]
pipeline:run
